### ISSUE TITLE
R version of ranger package exiting after thowing the error

### ISSUE THREAD

**Issue by ncdingari** [imbs-hl/ranger#209]
I am getting this error
`terminate called after throwing an instance of std::bad_alloc`

My code looks something like this

```
survival_model <- ranger(survival_formula,
                         data = trainData,  
                         seed = 2017,
                         importance = 'permutation',
                         mtry = 2,
                         verbose = T,
                         num.trees = 1500,
                         write.forest=T,
                         # save.memory = T,
                         num.threads = 2)
```

**Comment by mnwright** [comment]
Thanks! Could you provide a reproducible example or details on the dataset?

**Comment by ncdingari** [comment]
I could not provide the data, it is working OK with less number of rows. I think it is related to memory issue when you use large dataset. I am using 7 features, but my data set has 2 million rows, precisely 1994974.

**Comment by aminevsaziz** [comment]
i too confirm that i work with 1.5 mil rows 1 & this issue is driving me crazy ...!!!

**Comment by mnwright** [comment]
Could you try to simulate a dataset producing the error? 
@ncdingari it seems you tried the `save.memory` parameter. No effect?

**Comment by mangos47** [comment]
Hi, 

I'm getting the same error when dataset size exceeds a certain number, and it is not really a very large dataset (randomForest package, xgboost, h2o, J48 from Rweka, C50 all support much larger size of similar structured data).

below is a simulated example which will throw the memory allocation error on my machine:
----------------------------------------------------------------
library(ranger)
dat=data.frame(
  v1=sample(letters,1e6,replace = T),
  v2=sample(LETTERS,1e6,replace = T),
  v3=rnorm(1e6),
  v4=runif(1e6),
  v5=sample(1:50,1e6,replace = T),
  v6=factor(sample(1:6,1e6,replace = T))
)
t1=system.time({ rgr_fit1=ranger(v6~.,data = dat) })
t2=system.time({ rgr_fit2=ranger(dependent.variable.name="v6",data = dat) })
----------------------------------------------------------------
The error is:
Error: std::bad_alloc Ranger will EXIT now.
Error in ranger(v6 ~ ., data = dat) : User interrupt or internal error. 
Timing stopped at: 0.91 0.03 0.94

When the error happens, OS (windows 10) will also pop-up an error box saying memory low problem, although in task monitor there won't be any memory spike shown up.
Both the formula interface and explicitly specifying "dependent.variable.name" yield the same issue.

My R version():
platform       x86_64-w64-mingw32          
arch           x86_64                      
os             mingw32                     
system         x86_64, mingw32             
status                                     
major          3                           
minor          4.2                         
year           2017                        
month          09                          
day            28                          
svn rev        73368                       
language       R                           
version.string R version 3.4.2 (2017-09-28)
nickname       Short Summer

same issue happens both when running in native R console, and running in Rstudio (1.0.153)

Sometimes the error seems to also corrupt OS and will make my computer hang. My computer is running windows 10 x64 Enterprise. spec: i7-7600U, 16GB ram, 500GB SSD(most likely irrelevant)

Setting save.memory=T will prevent the error, but the training speed slows down drastically (about 10x slower when I experimented on smaller dataset). the largest working dataset is around 300k rows and takes ~15mins to train, my intended full training set is ~ 2M rows, and if I were forced to use save.memory=T, I suspect it will take 10+ hours ;(

btw: when testing on smaller sample size for which ranger does work (up to 300k rows with 27 categorical variables and 5 numerical variables for multi-class classification problem), ranger's accuracy is very competitive against xgboost and h2o's randomforest, much better than randomForest and a few other tree-based model. I really hope the memory issue can be resolved as it is a very good performing package when it works.

Thanks!

**Comment by mnwright** [comment]
Thanks for the reproducible example. I can confirm the problem and I'll try to find the cause. It seems trees with many nodes take too much space.

**Comment by mnwright** [comment]
I've adde some `shrink_to_fit()` calls to free memory earlier. Better now?

**Comment by mangos47** [comment]
I tried with GitHub version ( devtools::install_github("imbs-hl/ranger") ) this morning, and still have the same memory allocation error. The error shows up before any on-screen progress text, probably building the first tree. Thanks

**Comment by mnwright** [comment]
Strange. Your simulated example is working fine for me on a 8GB Mac and a 4GB Ubuntu VM. I'll have to try on a Windows machine.

**Comment by mangos47** [comment]
Thanks for the update!

I tried the latest GitHub version (0.8.3, date 2017-10-30), the memory error is still there. I then tried the same script on another machine which is running windows server 2012 R2, with 4 core AMD Opteron and 12 GB ram, same error. Maybe it is a windows specific issue but I don't have Linux or mac to test at the moment. I'll try test another windows 7 machine I have later.

**Comment by mangos47** [comment]
So I tried ranger on an Ubuntu VM inside my windows 10, and it has no issue with larger dataset that would otherwise crash in windows 10. The VM is only allocated half of total physical memory. So it is most likely a windows OS specific issue.

**Comment by mnwright** [comment]
I tried on a Windows machine and can confirm the problem. The error appears almost immediately after calling ranger. 

It runs fine on a 1GB Ubuntu VM (AWS t2.micro) and is crashing on a 8GB Windows machine.

**Comment by mangos47** [comment]
Any update on this issue?

**Comment by mnwright** [comment]
I tried again on another Windows machine with Windows Server 2008 and it's running fine there. I'll have to debug on my computer at work when I'm back there in January. 

Same problem with `num.threads=1`?

**Comment by mnwright** [comment]
Try again please. We were allocating a lot of memory at the same time. It seems Win 7 doesn't like this (but Windows Server 2008 does...).

**Comment by mangos47** [comment]
The latest version seems to have solved the issue. I've tried the previous mock code with 1e7 rows and 5 features without memory error anymore, though it takes a very long time on default settings.

Thanks for the great package!

**Comment by arturakb** [comment]
I have a similar problem with the following code on Linux, CentOS 7:

```
rm(list = ls())

library(parallel)
library(data.table)
library(ranger)

print(detectCores())

n <- 35000
p <- 30

m <- matrix(NA_real_, nrow = n, ncol = p)
colnames(m) <- paste0('Var', 1:ncol(m))
rownames(m) <- paste0('ID', 1:nrow(m))

m[, 1:10] <- rnorm(10 * n)
m[, 11:20] <- sample(c(1, 0), 10 * n, replace = TRUE)

dt <- as.data.table(m)

for (j in 21:p) {
  levs <- letters[sample(3:length(letters), sample(3:5, 1))]
  dt[, (j) := as.factor(sample(levs, .N, replace = TRUE))]
}

dt[, Status := sample(c(0, 1), .N, replace = TRUE)]
dt[, Survival_time := rlnorm(.N)]

system.time(
  mod <- ranger(Surv(Survival_time, Status) ~ .,
                data = dt,
                num.trees = 1000,
                mtry = 10,
                splitrule = 'logrank',
                num.threads = 10,
                min.node.size = 100)
)
```

**Comment by mnwright** [comment]
The problem is that you have 35000 unique time points and the CHF is evaluated at each unique time points. We don't have an option to restrict the number of unique time points yet (see #410), but an easy solution is to manually restrict them, e.g. by using 

`dt[, Survival_time := rlnorm(.N)]`

in your example. See also #202.
